dimod/bqm/adjmapbqm.pyx

```diff
@@ -125,69 +141,175 @@ cdef class AdjMapBQM:
+
+        Returns:
+            hashable: The label of the added variable.
+
+        Raises:
+            TypeError: If the label is not hashable.
+
+        Examples:
+
+            >>> bqm = dimod.AdjMapBQM()
+            >>> bqm.add_variable()
+            0
+            >>> bqm.add_variable('a')
+            'a'
+            >>> bqm.add_variable()
+            2
+
+            >>> bqm = dimod.AdjMapBQM()
+            >>> bqm.add_variable(1)
+            1
+            >>> bqm.add_variable()  # 1 is taken
+            0
+            >>> bqm.add_variable()
+            2
+
+        """
 
         if label is None:
             # if nothing has been overwritten we can go ahead and exit here
             # this is not necessary but good for performance
-            if not self.label_to_idx:
+            if not self._label_to_idx:
                 return add_variable(self.adj_)
 
             label = num_variables(self.adj_)
 
+            if self.has_variable(label):
+                # if the integer label already is taken, there must be a missing
+                # smaller integer we can use
+                for v in range(label):
+                    if not self.has_variable(v):
+                        break
+                label = v
 
-        # don't create variables that already exist
-        if self.has_variable(label):
-            return label
-
-        cdef object v = add_variable(self.adj_)
+        else:
+            try:
+                self.label_to_idx(label)
+            except ValueError:
+                pass
+            else:
+                # it exists already
+                return label
 
-        if v != label:
-            self.label_to_idx[label] = v
-            self.idx_to_label[v] = label
+        cdef object vi = add_variable(self.adj_)
+        if vi != label:
+            self._label_to_idx[label] = vi
+            self._idx_to_label[vi] = label
 
         return label
 
     def has_variable(self, object v):
-        if v in self.label_to_idx:
-            return True
-        return (v in range(num_variables(self.adj_)) and  # handles non-ints
-                v not in self.idx_to_label)  # not overwritten
+        """Return True if the binary quadratic model contains variable v.
+
+        Args:
+            v (hashable):
+                A variable in the binary quadratic model.
+
+        """
+        try:
+            self.label_to_idx(v)
+        except ValueError:
+            return False
+        return True
 
     def iter_variables(self):
         cdef object v
         for v in range(num_variables(self.adj_)):
-            yield self.idx_to_label.get(v, v)
+            yield self._idx_to_label.get(v, v)
 
     def pop_variable(self):
+        """Remove a variable from the binary quadratic model.
+
+        Returns:
+            hashable: The last variable added to the binary quadratic model.
+
+        Raises:
+            ValueError: If the binary quadratic model is empty.
```

Comment:
As commented on in https://github.com/dwavesystems/dimod/pull/523#discussion_r319322082, raising a `ValueError` here might not make sense?